java/build/bazel/tests/integration/BazelCommand.java

```diff
@@ -0,0 +1,148 @@
+  public List<String> outputLines() {
+    return delegate.getOutputLines();
+  }
+
+  /**
+   * Returns a description/summary of the command arguments, standard error and output stream. This
+   * method can be used for troubleshooting and error reporting.
+   */
+  public String toString() {
+    String description =
+        "BAZEL COMMAND: "
+            + args
+            + "\n"
+            + "EXIT CODE: "
+            + exitCode
+            + "\n"
+            + "STDOUT:\n    "
+            + String.join("\n    ", outputLines())
+            + "\n"
+            + "STDERR:\n    "
+            + String.join("\n    ", errorLines())
+            + "\n";
+
+    return description;
+  }
+
+  static class Builder {
+    private final WorkspaceDriver driver;
+    private final Path outputUserRoot;
+    private final List<String> args;
+    private Path bazelrcFile = null;
+    private Map<String, String> environment = new HashMap<>();
+    private Path workingDirectory = Paths.get("");
+
+    Builder(WorkspaceDriver driver, Path outputUserRoot, List<String> args) {
+      this.driver = driver;
+      this.outputUserRoot = outputUserRoot;
+      this.args = Collections.unmodifiableList(args);
+    }
+
+    /** sets the working directory relative to workspace path */
+    public Builder inWorkingDirectory(Path dirRelativeToWorkspaceRoot) {
+      this.workingDirectory = dirRelativeToWorkspaceRoot;
+      return this;
+    }
+
+    /** sets the bazelrc file path */
+    public Builder withBazelrcFile(Path bazelrcFile) {
+      this.bazelrcFile = bazelrcFile;
+      return this;
+    }
+
+    /** adds environment variable to execution runtime */
+    public Builder withEnvironmentVariable(String key, String value) {
+      this.environment.put(key, value);
+      return this;
+    }
+
+    private Command build() throws IOException {
+      String bazelRcPath =
+          bazelrcFile == null
+              ? "/dev/null"
+              : driver.currentWorkspace().resolve(bazelrcFile).toString();
+
+      List<String> command =
+          Stream.concat(
+                  Stream.of(
+                      driver.bazelBinPath().toString(),
+                      "--output_user_root=" + outputUserRoot,
+                      "--nomaster_bazelrc",
+                      "--max_idle_secs=10",
+                      "--bazelrc=" + bazelRcPath),
+                  args.stream())
+              .collect(Collectors.toList());
+
+      Path relativeToWorkspaceFullPath = driver.currentWorkspace().resolve(workingDirectory);
+
+      return Command.builder()
+          .setDirectory(relativeToWorkspaceFullPath)
+          .addArguments(command)
+          .withEnvironment(environment)
+          .build();
+    }
+
+    public BazelCommand run() throws IOException, InterruptedException {
+      Command cmd = build();
+      return new BazelCommand(cmd, args, cmd.run());
+    }
+
+    public BazelCommand mustRun() throws IOException, InterruptedException {
```

Comment:
the phrasing `mustRun` confused me when I read the tests. WDYT about `mustRunSuccessfully`? longer but more revealing IMHO

Reply:
That's derived from a convention in Golang.  That might be a bit foreign here, so I'll go with your suggestion.